### PR TITLE
Flver parser mesh

### LINKED ISSUE #14: Create starter DS FLVER Parser with ability to parse and output meshes from Dark Souls .flver files.

### PR DESCRIPTION
**Pull request by augenblick**
A first 'working' version  of the flver_extractor module.  Models extract and obj files can be created, but there are some face and vert normal issues.

Closes #14

### CODE REVIEW

File: ds_import/binary_reader.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
+import os
+import struct
+import binascii
```
Comment (aaronaskew) [inline ds_import/binary_reader.py]: Don't think os or binascii get used. Remove for now?

File: ds_import/binary_reader.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+class BinaryReader:
+
+    _loaded_file_path = ""
+    _endianness = ""
+    _reader_file = None
+    _reader_position = None
+
+    # TODO: allow for big-endian byte order?
```
Comment (aaronaskew) [inline ds_import/binary_reader.py:11]: I see some code below for big endian order. Has this TODO been TODONE?

File: ds_import/binary_reader.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            struct_fmt = "<" + struct_fmt
+
+        fmt_size = struct.calcsize(struct_fmt)
+        return_struct = struct.unpack(struct_fmt, self.reader_file.read(fmt_size))
+
+        self._reader_position += fmt_size
+        if len(return_struct) == 1:
+            return return_struct[0]     # if returning a single value, no need for a tuple
+        return return_struct
+        # TODO: return in a format besides tuple?
```
Comment (aaronaskew) [inline ds_import/binary_reader.py]: Just curious why tuple might be less preferred.

File: ds_import/binary_reader.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        self._reader_position += 1
+        return self.reader_file.read(1)
+
+
+    # returns an 8-bit signed integer from the current file position, and moves the position 8-bits forward
+    def get_int8(self):
+        return self.get_struct("b")
+
+
+    # returns an 8-bit signed integer from the current file position, and moves the position 8-bits forward
```
Comment (aaronaskew) [inline ds_import/binary_reader.py]: unsigned

File: ds_import/binary_reader.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    def get_u_int8(self):
+        self._reader_position += 1
+        return self.get_struct("B")
+
+
+    # returns a 32-bit signed integer from the current file position, and moves the position 32-bits forward
+    def get_int32(self):
+        self._reader_position += 4
+        return self.get_struct("i")
+        # return int.from_bytes(self.reader_file.read(4), byteorder=self.endianness)
```
Comment (aaronaskew) [inline ds_import/binary_reader.py]: this comment line delete-able?

File: ds_import/binary_reader.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+
+    # returns a 32-bit signed integer from the current file position, and moves the position 32-bits forward
```
Comment (aaronaskew) [inline ds_import/binary_reader.py]: unsigned

File: ds_import/flver_extractor.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+            self.__material_parameters = \
+                self.__obtain_info(reader, "IIffIIII", self.__metadata.material_param_count, self.material_parameters)
+
+            self.__vertex_struct_formats = self.__define_vertset_formats()
+
+    # check to see whether the file is a readable format
+    def __is_ds1_flver(self):
+
+        if not os.path.isfile(self.__flver_file_path):
+            raise FileNotFoundError("The file '{filepath}' could not be found".format(filepath=self.__flver_file_path))
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: I feel like this check should go in the BinaryReader class as an effort to try to keep all the OS/file type stuff in one place, and the extractor only has to pass strings of file locations.

File: ds_import/flver_extractor.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+                self.__obtain_info(reader, "IIII", self.__metadata.vertex_descr_count, self.vertex_description_info)
+
+            # collect material parameters
+            self.__material_parameters = \
+                self.__obtain_info(reader, "IIffIIII", self.__metadata.material_param_count, self.material_parameters)
+
+            self.__vertex_struct_formats = self.__define_vertset_formats()
+
+    # check to see whether the file is a readable format
+    def __is_ds1_flver(self):
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py:286]: I get what this is doing, but a comment spelling out the DS1 file format criteria in plain english would be nice in the docstring.

**Review by aaronaskew — CHANGES_REQUESTED** [review]
Made a few notes on things. Still have a good bit of flvr_extractor to review, but leaving what I have for now.

File: ds_import/binary_reader.py (hunk 1 above)
Comment (augenblick) [reply ds_import/binary_reader.py]: yeah, will remove.

File: ds_import/binary_reader.py (hunk 2 above)
Comment (augenblick) [reply ds_import/binary_reader.py:11]: it's partly supported in a couple of places, but not fully supported throughout.

Comment (augenblick) [reply ds_import/binary_reader.py:11]: I'll leave this in for now, but we should revisit and decide if this is worth pursuing or not.

File: ds_import/binary_reader.py (hunk 3 above)
Comment (augenblick) [reply ds_import/binary_reader.py]: tuple seem to be working fine-- '"ll remove this comment.

File: ds_import/binary_reader.py (hunk 4 above)
Comment (augenblick) [reply ds_import/binary_reader.py]: I'll correct this.

File: ds_import/binary_reader.py (hunk 5 above)
Comment (augenblick) [reply ds_import/binary_reader.py]: yep!  I'll delete it.

File: ds_import/binary_reader.py (hunk 6 above)
Comment (augenblick) [reply ds_import/binary_reader.py]: yep! I'll correct it.

File: ds_import/flver_extractor.py (hunk 7 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: Yeah, good point, and I agree.  I'll move this into `BinaryReader`

File: ds_import/flver_extractor.py (hunk 8 above)
Comment (augenblick) [reply ds_import/flver_extractor.py:286]: I agree this is a good place for some docstring explanation-- I'll add it.

File: ds_import/model.py [hunk 9]
```diff
@@ -15,3 +15,5 @@ def __init__(self, meshes=None):
         for m in meshes:
             if not isinstance(m, Mesh):
                 raise TypeError("Objects of type Mesh expected, however type {} was passed".format(type(m)))
+
+        self.meshes = meshes
```
Comment (aaronaskew) [inline ds_import/model.py]: Deleting this, as model.py has been thoroughly reworked.

File: ds_import/flver_extractor.py [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+        # bounding_box_2 (vec3)
+        # unknown_3 (u int32)
+        # unknown_4 (u int32)
+        # unknown_5 (u int32)
+        # unknown_6 (u int32)
+        faceset_count = 0
+        vertex_descr_count = 0
+        material_param_count = 0
+
+        vert_pointers = None
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: What is vert_pointers?

File: ds_import/flver_extractor.py [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+
+
+        def __init__(self, flver_file):
+
+            with BinaryReader(flver_file) as reader:
+                reader.seek(12)
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: Please explain this seek via comment.

File: ds_import/flver_extractor.py [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+                self.data_offset = reader.get_u_int32()
+                self.data_size = reader.get_u_int32()
+                self.hitbox_count = reader.get_u_int32()
+                self.material_count = reader.get_u_int32()
+                self.bone_count = reader.get_u_int32()
+                self.mesh_count = reader.get_u_int32()
+                self.vertex_info_count = reader.get_u_int32()
+                reader.seek(80)
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: Please explain this seek via comment.

File: ds_import/flver_extractor.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+    __flver_file_path = ""
+    __metadata = None                   # flver file metadata-- data offset, number of meshes, bones, etc.
+
+    __materials_info = None             # per material: offsets to material names, material paths, parameter count, etc.
+    __mesh_info = None                  # per mesh: is_dynamic, material_index, bone_index, etc.
+    __faceset_info = None               # per faceset: information about each faceset buffer (indices count, offset to indices, etc.)
+    __vertex_info = None                # per vertex set: vertex_struct_formats_index, data size per vertex, vertices offset, vertex_count, etc.
+    __vertex_description_info = None    # per vertex description: count of datapoints in this description, offset to start of description
+    __material_parameters = None        # per material?: offsets to material/texture filenames(?)
+    __vertex_struct_formats = None      # list of ordered dictionaries describing the vertex data layouts
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: These should be instance variables instead of class variables. I only recently learned about this. Coming from a Java background, I think of class variables like static variables. https://docs.python.org/3/tutorial/classes.html#class-and-instance-variables

Basically all of these should be declared as `self.__variable_name` within the `__init__` method. 

We currently are only using one FlverExtractor instance at a time, but we may want to have more at some point.

File: ds_import/flver_extractor.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+from collections import namedtuple
+from face import Face
+from face_set import FaceSet
+from model import Model
+from vertex import Vertex
+from mesh import Mesh
+from vector2 import Vector2
+from vector3 import Vector3
+from binary_reader import BinaryReader
+from errors import UnreadableFormatError
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: Prepend `ds_import.` to all module names to avoid relative import issues later.

File: ds_import/obj_writer.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
+from model import Model
```
Comment (aaronaskew) [inline ds_import/obj_writer.py]: Prepend `ds_import.` to `model`

File: ds_import/flver_extractor.py [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+    mesh_info = namedtuple("mesh_info", ["is_dynamic", "material_index", "unknown1", "unknown2", "bone_index", "bone_index_count", "unknown3", "bone_indices_offset",
+                                        "faceset_count", "faceset_index_offset", "vertex_info_count", "vertex_info_offset"])
+    faceset_info = namedtuple("faceset_info", ["flags", "topology", "cull_backfaces", "unknown1", "unknown2",
+                                               "index_count", "buffer_offset", "buffer_size", "unknown3", "unknown4", "unknown5"] )
+    vertex_info = namedtuple("vertex_info", ["unknown1", "vertex_struct_formats_index", "per_vertex_size", "vertex_count",
+                                             "unknown2", "unknown3", "buffer_size", "buffer_offset"])
+    vertex_description_info = namedtuple("vertex_description_info", ["datatype_count", "unknown1", "unknown2", "description_offset"])
+    material_parameters = namedtuple("material_parameters", ["name_offset1", "name_offset2", "unknown1", "unknown2", "unknown3", "unknown4", "unknown5", "unknown6"])
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: This seems like a good spot to document what all the dictionary keys actually are. It's not quite clear to me what each of the names mean.

File: ds_import/flver_extractor.py [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+        vert_data_dict = dict({"position":"fff",
+                              "bone index":"I",
+                              "normal":"BBBx",
+                              "vertex color":"BBBB",
+                              "bitangent":"BBBB",
+                              "diffuse/lightmap UV":"HHxxxx",
+                              "bone weight":"II",
+                              "diffuse UV":"HH",
+                              "unknown":"I"},
+                              )
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: Think this belongs at the top near the other class variables.

File: ds_import/flver_extractor.py [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+        Extracts the loaded Dark Souls flver file and returns as a Model
+
+        :return: A Dark Souls Model
+        """
+
+        if self.__flver_file_path == "":
+            raise Exception("A file must be loaded before extraction.  Use FlverExtractor.load_file(flver_filepath)")
+
+
+        # TODO: fix bug occurring when there is more than one faceset per mesh.  There is a "faceset_count" parameter stored in self.__mesh_info that probably should be referenced
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: Create an issue?

File: ds_import/flver_extractor.py [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+            for vi in self.__vertex_info:
+
+                current_vertex_format = ""
+                reader.seek(vi.buffer_offset + self.__metadata.data_offset)
+                for datatype in self.__vertex_struct_formats[vi.vertex_struct_formats_index]:
+                    current_vertex_format += vert_data_dict.get(datatype)
+                vert_list = []
+                for vertex in range(0,vi.vertex_count):
+                    vert_data = reader.get_struct(current_vertex_format)
+                    new_vert = Vertex()
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: Moving `new_vert = Vertex()` just before the for loop could really help performance/memory usage probably, as long as you still clear it's member variables to `None` as you're already doing. That could potentially be a lot of Vertex objects to allocate and do garbage collection on.

File: ds_import/flver_extractor.py [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+                            data_index += 2
+
+                        elif data_type == "unknown":
+                            # TODO: handle "unknown" case
+                            data_index += 1
+                            pass
+
+                    vert_list.append(new_vert)
+                new_mesh = Mesh(vert_list)
+                mesh_list.append(new_mesh)
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: Think it would be more efficient to do `mesh_list.append(Mesh(vert_list))`. Due to python passing by assignment, the above is double allocating Mesh instances.

File: ds_import/flver_extractor.py (hunk 16 above)
Comment (aaronaskew) [reply ds_import/flver_extractor.py]: Also, I think `_info` is confusing, as these don't really hold the information, but rather describe how the info is organized. 

Maybe `mesh_info_mapping` or `mesh_info_structure`? Ideas?

File: ds_import/flver_extractor.py [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+                    faceslist.append(f3)
+                else:
+                    faceslist.append(f1)
+                    faceslist.append(f3)
+                    faceslist.append(f2)
+            counter = counter + 1
+            f1 = f2
+            f2 = f3
+
+        return faceslist
```
Comment (aaronaskew) [inline ds_import/flver_extractor.py]: This took me down a weird rabbit hole. I think because of the naming. This isn't sorting faces, it's sorting the _vertices_ of the faces, correct? Also the comment `# sorts a given faceset into an order that Blender can use` threw me off because I had already created a FaceSet class that does something completely different. 

What I think is happening with faces, and please correct me if I'm wrong, is that there's a long list of ints being read that are the indices of the vert_list. Every 3 vertex indices are then parsed into a Face each. The sort is needed to ensure vertex orders expected by blender. 

The algorithm appears to:
* for all vertex indices
  * reverse the order used for the last vertex
  * create a set of 3 vertex indices
  * ensure it does not contain duplicate vertex indices (if there is a duplicate, it is skipped)
  * append them to the list. 

The returned list is shorter than the original.

My suggested naming changes below. 


```suggestion
    # sorts a given list of vertex indices defining a set of faces into an order that Blender can use
    def __sort_faces_vertex_list(self, faces_vertex_list):

        # TODO: figure out why this works
        sorted_faces_vertex_list = []
        StartDirection = -1
        face_vertex1 = faces_vertex_list[0]
        face_vertex2 = faces_vertex_list[1]
        FaceDirection = StartDirection

        counter = 2
        while counter < len(faces_vertex_list):
            face_vertex3 = faces_vertex_list[counter]
            
            # For each vertex in the list, reverse the order of the vertices, 
            # effectively reversing the normal of the face.
            FaceDirection *= -1

            # If any of the three vertices in this set are duplicates, 
            # do not append, effectively skipping one vertex.
            if (face_vertex1 != face_vertex2) and (face_vertex2 != face_vertex3) and (face_vertex3 != face_vertex1):
                if FaceDirection > 0:
                    sorted_faces_vertex_list.append(face_vertex1)
                    sorted_faces_vertex_list.append(face_vertex2)
                    sorted_faces_vertex_list.append(face_vertex3)
                else:
                    sorted_faces_vertex_list.append(face_vertex1)
                    sorted_faces_vertex_list.append(face_vertex3)
                    sorted_faces_vertex_list.append(face_vertex2)
                    
            counter = counter + 1
            
            # Shift the set of vertices down the list by one
            face_vertex1 = face_vertex2
            face_vertex2 = face_vertex3

        return sorted_faces_vertex_list
```

**Review by aaronaskew — CHANGES_REQUESTED** [review]
Ok, got through another chunk. Not quite done.

File: ds_import/flver_extractor.py (hunk 8 above)
Comment (aaronaskew) [reply ds_import/flver_extractor.py:286]: Awesome.

File: ds_import/flver_extractor.py (hunk 13 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: Ah!  Thanks for pointing that out--  I'll move these to the constructor.

It's weird though-- most of the documentation mentions the setup you've described, but the action of the interpreters I've used seems more nuanced, like it will create underlying instance variables on the fly if a class variable is referenced using "self.class_variable".  (This is how I've been referencing them throughout the project)  I haven't really looked deeply into it, but I think it's interesting.

Try running this if you get a second:

```
class TestClass:

	test_variable = -1


	def __init__(self, value):

		self.test_variable = value


	def print_test_variable(self):
		print("(" + str(self) + ") called as self.test_variable: " + str(self.test_variable))
		print("(" + str(self) + ") called as TestClass.test_variable: " + str(TestClass.test_variable))

	@staticmethod
	def print_static_test_variable():
		print("called from static method: " + str(TestClass.test_variable))


tester_1 = TestClass(422)
tester_2 = TestClass(1138)

tester_1.print_test_variable()
tester_2.print_test_variable()
TestClass.print_static_test_variable()
```

File: ds_import/flver_extractor.py (hunk 10 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: That was left in by mistake.  I started to put some of the namedstructs (faceset_info, vertex_info, etc.) in the metadata inner class.  vert_pointers was what I had considered calling one of these namedstructs.

I'll remove it.

File: ds_import/flver_extractor.py (hunk 11 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: done.

File: ds_import/flver_extractor.py (hunk 12 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: done.

File: ds_import/flver_extractor.py (hunk 18 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: We (probably) wouldn't need this unless we want to at some point support different LOD models.  

Shouldn't be too bad a fix, so I say let's go for it.  I'll make an issue.

File: ds_import/flver_extractor.py (hunk 13 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: On second thought, nevermind.  After some _very keen_ observation I see where I'm getting turned around.  Clearly I'm explicitly declaring the instance variable in the constructor of the above example.

File: ds_import/flver_extractor.py (hunk 18 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: It's not confirmed that this class is the source of the a bug.  The original comment is outdated, and there is definitely some bug that exists when multiple facesets exist, but I'm not convinced it's happening here rather than when a model is converted to an obj.  I'll create an issue to investigate this further.

File: ds_import/flver_extractor.py (hunk 13 above)
Comment (aaronaskew) [reply ds_import/flver_extractor.py]: I actually think the class variables are currently declared and assigned None, and then when you are referencing them as self.blah the instance variables are created on the fly, thus the class variable declarations are never used. 

I think as a matter of convention and clarity, it just makes sense to set up any class variables (static variables) at the top, and initialize instance "selfies" in the constructor. Sound good?

File: ds_import/flver_extractor.py (hunk 14 above)
Comment (aaronaskew) [reply ds_import/flver_extractor.py]: I think you already did this, yes?

File: ds_import/flver_extractor.py (hunk 18 above)
Comment (aaronaskew) [reply ds_import/flver_extractor.py]: Cool. remove the outdated todo?

File: ds_import/flver_extractor.py (hunk 19 above)
Comment (aaronaskew) [reply ds_import/flver_extractor.py]: Just realized that the Mesh class, along with the rest, will raise a TypeError if you assign None to the member variables. I can create a .reset() method to re-initialize the instance instead. Sound good?

File: ds_import/flver_extractor.py (hunk 13 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: Yeah, I'll bet the class variables are sitting around unused.  It looks like the instance variables are being declared as they're defined throughout the code.  Quite a few, for instance, in the load_file method.

I totally agree with your thinking on this.  It makes sense to create instance variables in the constructor.

File: ds_import/flver_extractor.py (hunk 19 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: yeah, sounds good!
And thanks for pointing out this optimization too.  I'll make that change.

File: ds_import/flver_extractor.py (hunk 20 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: good idea.  I'm on it.

File: ds_import/flver_extractor.py (hunk 21 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: Yeah, you're right about "faces" in this case really being indices into the vertex list.

That looks like a solid breakdown for how this method is converting the native flver ordering into a blender-ready format.  What I'm very curious about is why it's stored that way in the Dark Souls data.  Seems wasteful, but there must be an explanation for it.  I may spend some time looking at it for fun later.

Comment (aaronaskew) [reply ds_import/flver_extractor.py]: "fun"

File: ds_import/flver_extractor.py (hunk 14 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: Yeah, all done with this one.

File: ds_import/flver_extractor.py (hunk 16 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: I've given this a go-- but it still needs work.  It bloats the script a bit, but I think it's worth having the information close to the code, at least for now.  That said,  I'll include the information in the wiki when I get a chance.

Comment (augenblick) [reply ds_import/flver_extractor.py]: I've made the changes, but have run into some merge conflicts and will have to sort it out tomorrow.

**Comment by augenblick** [comment]
I've made a pass at all requested changes.

File: ds_import/flver_extractor.py (hunk 16 above)
Comment (augenblick) [reply ds_import/flver_extractor.py]: I'm down for updating the names of these namedstructs.  Just need to figure out what names make the most sense.

Comment (aaronaskew) [reply ds_import/flver_extractor.py]: Created issue #42.

**Review by aaronaskew — APPROVED** [review]
